### PR TITLE
GitHub Organization Restriction for APIView

### PR DESCRIPTION
**Pull request by mccoyp**
The primary change in this pull request is the addition of GitHub organization-based authorization. Currently, users who are not publicly visible members of the Azure organization won't have site access - a change to make each uploaded assembly independently organization-restricted will be underway soon.

Also, a number of bugs have been resolved:
- Namespaces no longer render as nested
- Empty namespaces are hidden
- Extension methods' first parameters are preceded by the `this` keyword
- Code model types have had suffixes changed to Pascal case (`-APIV` to `-ApiView`), and the `APIView` namespace has been changed to `ApiView`

### CODE REVIEW

**Comment by pakrym** [comment]
Never do mass renames and meaningful changes in the same PR. Makes it very hard to review.

File: src/dotnet/APIView/APIViewWeb/Startup.cs [hunk 1]
```diff
@@ -81,11 +86,42 @@ public void ConfigureServices(IServiceCollection services)
...
+                            }
+
+                            context.Identity.AddClaim(new Claim("urn:github:orgs", orgNames.ToString()));
                         }
                     };
                 });
+
+            services.AddAuthorization(options => {
+                options.AddPolicy("RequireOrgAzure", policy => 
+                    policy.RequireOrganizationRequirement("Azure"));
```
Comment (pakrym) [inline src/dotnet/APIView/APIViewWeb/Startup.cs]: Put this into configuration.

File: src/dotnet/APIView/APIViewWeb/ExtensionMethods.cs [hunk 2]
```diff
@@ -9,4 +12,41 @@ public static string GetGitHubLogin(this ClaimsPrincipal user)
             return user.FindFirst(c => c.Type == "urn:github:login")?.Value;
         }
     }
+    public class OrganizationRequirement: IAuthorizationRequirement
```
Comment (pakrym) [inline src/dotnet/APIView/APIViewWeb/ExtensionMethods.cs]: 1 type per file

File: src/dotnet/APIView/APIViewWeb/ExtensionMethods.cs [hunk 3]
```diff
@@ -9,4 +12,41 @@ public static string GetGitHubLogin(this ClaimsPrincipal user)
...
+    {
+        public string Org { get; set; }
+        public OrganizationRequirement(string org)
+        {
+            Org = org;
```
Comment (pakrym) [inline src/dotnet/APIView/APIViewWeb/ExtensionMethods.cs]: Fix the name

File: src/dotnet/APIView/APIViewWeb/ExtensionMethods.cs [hunk 4]
```diff
@@ -9,4 +12,41 @@ public static string GetGitHubLogin(this ClaimsPrincipal user)
...
+        }
+    }
+
+    public class OrganizationRequirementHandler : IAuthorizationHandler
+    {
+        public Task HandleAsync(AuthorizationHandlerContext context)
+        {
+            foreach (var r in context.Requirements)
```
Comment (pakrym) [inline src/dotnet/APIView/APIViewWeb/ExtensionMethods.cs]: variable names

File: src/dotnet/APIView/APIViewWeb/ExtensionMethods.cs [hunk 5]
```diff
@@ -9,4 +12,41 @@ public static string GetGitHubLogin(this ClaimsPrincipal user)
...
+            {
+                if (r is OrganizationRequirement o)
+                {
+                    try
+                    {
+                        if (context.User.FindFirst("urn:github:orgs").Value.Split(",").Contains(o.Org))
```
Comment (pakrym) [inline src/dotnet/APIView/APIViewWeb/ExtensionMethods.cs]: Would FindFirst return null if a claim doesn't exist?

File: src/dotnet/APIView/APIView/MethodApiView.cs [hunk 6]
```diff
@@ -56,25 +57,29 @@ public MethodAPIV(IMethodSymbol symbol)
...
+            List<ParameterApiView> parameters = new List<ParameterApiView>();
 
             foreach (AttributeData attribute in symbol.GetAttributes())
             {
-                if (attribute.AttributeClass.DeclaredAccessibility == Microsoft.CodeAnalysis.Accessibility.Public || 
-                    attribute.AttributeClass.DeclaredAccessibility == Microsoft.CodeAnalysis.Accessibility.Protected)
-                attributes.Add(new AttributeAPIV(attribute, this.Id));
+                if ((attribute.AttributeClass.DeclaredAccessibility == Microsoft.CodeAnalysis.Accessibility.Public || 
+                    attribute.AttributeClass.DeclaredAccessibility == Microsoft.CodeAnalysis.Accessibility.Protected) &&
+                    attribute.AttributeClass.Name != "AsyncStateMachineAttribute" &&
```
Comment (pakrym) [inline src/dotnet/APIView/APIView/MethodApiView.cs]: Define a list of ignore attribute names

File: src/dotnet/APIView/APIViewWeb/ExtensionMethods.cs (hunk 5 above)
Comment (mccoyp) [reply src/dotnet/APIView/APIViewWeb/ExtensionMethods.cs]: It does - I'm not sure if there's a better way to go about this check. I added a requirement in the `RequireOrganizationRequirement` for an organizations claim, but when the site initially spins up, this line in the code is triggered and returns null.

File: src/dotnet/APIView/APIViewWeb/ExtensionMethods.cs (hunk 2 above)
Comment (mccoyp) [reply src/dotnet/APIView/APIViewWeb/ExtensionMethods.cs]: Should I put both extension methods in a generic extension method class in this file?

File: src/dotnet/APIView/APIViewWeb/ExtensionMethods.cs (hunk 5 above)
Comment (pakrym) [reply src/dotnet/APIView/APIViewWeb/ExtensionMethods.cs]: Check claim for null first

File: src/dotnet/APIView/APIViewWeb/ExtensionMethods.cs (hunk 2 above)
Comment (pakrym) [reply src/dotnet/APIView/APIViewWeb/ExtensionMethods.cs]: Nah, they are unrelated. But you can remove this extension method altogether, it has only one usage.

File: src/dotnet/APIView/APIView/MethodApiView.cs [hunk 7]
```diff
@@ -64,14 +64,16 @@ public MethodApiView(IMethodSymbol symbol)
             List<TypeParameterApiView> typeParameters = new List<TypeParameterApiView>();
             List<ParameterApiView> parameters = new List<ParameterApiView>();
 
+            List<string> ignoredAttributeNames = new List<string>() { "AsyncStateMachineAttribute", "DebuggerStepThroughAttribute" };
```
Comment (pakrym) [inline src/dotnet/APIView/APIView/MethodApiView.cs]: static

**Pull request merged**